### ISSUE TITLE
Shapes Editor errors

### ISSUE THREAD

**Issue by triksterr** [litdev1/SB-IDE#7]
1) When I insert a Browser Shape I get an error: missing address - the 3rd parameter cannot be an empty string.

2) The changing of parameters of shapes case an incorrect results: 
- parameters cannot have zero value: for example, I can't draw vertical or horisontal line.
- parameters cannot have integer value: all new values are noninteger digits with decimal points
- in code window parameters writed as digits with comma instead of decimal point.

**Comment by litdev1** [comment]
Hi,

The browser issue is just that the default website I added is now trying to run java scripts (not allowed for security outside a real browser).  I will use a different default site.

Please can you provide more details on the other issues - ie how to reproduce exactly.  I have issue trying to drag a line that has been set to zero size - but this is pretty specific and maybe not what you are talking about.

Also the non-integer, and comma deliminators I cant reproduce or know how to try to - it could be localisation? - some languages use , for decimal deliminator.

**Comment by triksterr** [comment]
> The browser issue is just that the default website I added is now trying to run java scripts (not allowed for security outside a real browser). I will use a different default site.

Ок. I think you can use your site. :)

> Also the non-integer, and comma deliminators I cant reproduce or know how to try to - it could be localisation? - some languages use , for decimal deliminator.

Probably. But SB uses only a dot - for all languages. I think you need to write a small parser that changes the comma (if any) to a point before inserted into the code.

**Comment by triksterr** [comment]
In this case, I cannot make the line vertical using the mouse. However, I can make it vertical by changing the value of X2 -> 0. But then if I try to use mouse to resize the line, the program crashes.
![Снимок](https://user-images.githubusercontent.com/48825903/77777752-0e2a1900-7061-11ea-9f0c-4ae7d413534d.JPG)

**Comment by triksterr** [comment]
In this case, I try to change the font size value (default 12) to 16 and I get - 15.92308!
![pic1](https://user-images.githubusercontent.com/48825903/77778385-f606c980-7061-11ea-9756-e18b21542ba2.JPG)

![pic2](https://user-images.githubusercontent.com/48825903/77778401-fc954100-7061-11ea-91c2-fab7c860228a.JPG)

![pic3](https://user-images.githubusercontent.com/48825903/77778417-01f28b80-7062-11ea-9162-e213e91ab195.JPG)

**Comment by litdev1** [comment]
"Probably. But SB uses only a dot - for all languages. I think you need to write a small parser that changes the comma (if any) to a point before inserted into the code."

To be totally clear, we mean numbers in the shape design (like font size) that are displayed with a comma.  When these are written to the Small Basic code window (the comma remains).  This is tricky for me to test since I have UK locale, but if this is the issue then I understand it.

The vertical line and font I should reproduce (sometime fonts are listed as e.g. 16, but in reality internally they are slightly different) and this may be the number that comes back from the font dialog in which case I might round it to the nearest 0.1?

**Comment by triksterr** [comment]
> To be totally clear, we mean numbers in the shape design (like font size) that are displayed with a comma. When these are written to the Small Basic code window (the comma remains). This is tricky for me to test since I have UK locale, but if this is the issue then I understand it.

Yes. It is a problem for foreign children who cannot understand "what's happend".

> The vertical line and font I should reproduce (sometime fonts are listed as e.g. 16, but in reality internally they are slightly different) and this may be the number that comes back from the font dialog in which case I might round it to the nearest 0.1?

Yes. About dimensions - It's Ok.

**Comment by litdev1** [comment]
I have checked in all the changes and uploaded a version to [http://litdev.co.uk/downloads/SB-Prime.zip](http://litdev.co.uk/downloads/SB-Prime.zip).

Extra ru strings are String119 to String 123.  Also added es strings to excel (as well as resx).

[localisation.xlsx](https://github.com/litdev1/SB-IDE/files/4395762/localisation.xlsx)

Extension manager is different dll and would need its own resources - leave that just now.
Message source keys I wanted to keep original and don't really affect understanding much
Color names are the colour names by SB code so I want to keep
The statusline bits are mostly created at run time and harder to translate

Hopefully fixed the line crash - division by 0 when width or height is zero

Recast the font size to nearest integer on return from FontDialog

Use Invariant Culture to write the decimal values to the SB code - hard for me to test with uk locale, but I think it should work?

Thanks for your interest to improve this tool.

**Comment by litdev1** [comment]
Ah just noticed your comment about colour - will have a look - I left a comment there.

Too many threads to this - perhaps close any completed

**Comment by triksterr** [comment]
This is Russian translations of new strings:

[localisation.xlsx](https://github.com/litdev1/SB-IDE/files/4396984/localisation.xlsx)

Please change ALL Russian strings to these!

**Comment by triksterr** [comment]
Ok. I will test this new version.

> Too many threads to this - perhaps close any completed

Each topic is for its own task. We will close them upon their completion. Ok?

**Comment by triksterr** [comment]
Today I updated SB-Prime to version 1.1.7. The problem with the straight line seems to have disappeared.

But I do not see the Russian language in the IDE interface. Russian language.
_Russian language appeared only after the second update ... :)_

And there was a problem with forms containing text.
![err17](https://user-images.githubusercontent.com/48825903/77933787-dc5ebf80-72b7-11ea-853d-59bde6ef0065.PNG)

**Comment by litdev1** [comment]
I can see why you need to update twice.  This is because the updater itself needs to be updated on the first update (to handle the language directories now present).  From here on it shuld need only one update - unless the updater needs changing.

I don't get the shapes editor issue, possibly because my localisation is uk - but your english error message suggests it is not the russian language?

Please confirm that the russian is working after second update and how to reproduce the shapes error.

**Comment by triksterr** [comment]
I tested it on version 1.1.7 after the first update - without Russian localization. After the second update, the problem still exists, but the message was in Russian.


![sss](https://user-images.githubusercontent.com/48825903/78011528-5e98c380-734c-11ea-9d79-0f44d871dc04.PNG)

**Comment by litdev1** [comment]
I can't reproduce this by just adding a button - my guess it is the height (mine is 24.5) - perhaps yours is 24,5 and the issue is the localisation of the comma in the numeral?

I am guessing a bit, and have uploaded a test that you can get by Advanced->Update.  I also added a try-catch with error message where I think the issue may be.

**Comment by triksterr** [comment]
Yes. The problem is - localization. In standard Russian settings, a decimal comma is used instead of a dot.
Still, it’s worth automatically turning the commas into dots in numbers.

![111](https://user-images.githubusercontent.com/48825903/78131617-092fe580-7424-11ea-8547-69c2a8dc51b3.PNG)

![222](https://user-images.githubusercontent.com/48825903/78131631-0e8d3000-7424-11ea-8be9-c87a5dd782b8.PNG)

**Comment by litdev1** [comment]
I was trying to do it 'the correct way' using localisations in the code rather than just replacing commas with decimal point.  We may get a different localisation where , is a deliminator between thousands or another.

However I can't test my code, so that is why I wanted you to try the version I uploaded.  If all else fails we can just give up and convert commas, but I would like to try a bit more since we got this far.

**Comment by triksterr** [comment]
Ok. I can test it every time. What about any other (more quick) way of communication between us. We can discuss it privately - by email. You can write me now.

**Comment by triksterr** [comment]
So. I used the SB-Prime-TEST.

After starting the Shapes Editor, I clicked "Button".

Then there were a few textboxes, and then I saw  "Fix String 26.5"

![12](https://user-images.githubusercontent.com/48825903/78511090-ac755780-77a2-11ea-8f60-d0a7a14ad733.PNG)

Then I saw an error message:

![13](https://user-images.githubusercontent.com/48825903/78511099-b8611980-77a2-11ea-8496-8442bf8f1779.PNG)

Then I saw "ShowCode 2"

![14](https://user-images.githubusercontent.com/48825903/78511103-c1ea8180-77a2-11ea-83a7-697201cdcbc0.PNG)

Do You need any other information?

**Comment by litdev1** [comment]
No that looks great I will work from that - it looks like it is getting the number in OK with a decimal, so not sure where the error may be.

**Comment by triksterr** [comment]
I think you need to find a place where the number converts to a string immediately before writing to the SB code.

**Comment by litdev1** [comment]
Sure - that is exacly what I am trying to do - however the message box shows that the string received has a decimal in it! - so I need to think again - the issue may be something else.
When I get some time after working today I will try to narrow it down and probablt create another test build.
If I knew where the issue was I would fix it.

**Comment by litdev1** [comment]
I have uploaded a new test version with more message box output - this is a slow proceedure but we should get there.

[http://litdev.co.uk/downloads/SB-Prime-TEST.zip](http://litdev.co.uk/downloads/SB-Prime-TEST.zip)

**Comment by triksterr** [comment]
Ok I will test it.

пн, 6 апр. 2020 г. в 19:45, litdev1 <notifications@github.com>:

**Comment by triksterr** [comment]
I did this: Mouse LeftClick -> Button

Then I saw these messages:

**AddShape1
ShowProperties 1
Fix Double 12
ShowProperties 2
Fix Double 52.9733333333333333
Fix Double 26.4833333333333333
Fix Double 100
Fix Double 100
ShowCode 1
ShowCode 2
Fix String 1 : 100
Fix String 2 : 100
Fix String 3 : 100
Fix String 1 : 100
Fix String 2 : 100
Fix String 3 : 100
ShowCode 3
Fix String 1 : 53
Fix String 2 : 53
Fix String 3 : 53
Fix String 1 : 26.5
Fix String 4 : 26.5**

Then I saw this error message:

**Error Message: "Входная строка имела неверный формат"**

And then I saw these messages:

**ShowCode 4
ShowCode 5
ShowCode 6**

The last screen was:
![Снимок](https://user-images.githubusercontent.com/48825903/78808940-d90db700-79ce-11ea-87b4-4beb21a7e8ee.PNG)

**Comment by litdev1** [comment]
Thanks that's great.

Interesting - it actually looks like its failing to convert string with decimal "26.5" to a number because it is expecting the string to be in local localisation with a comma.

I have uploaded another test build hopefully fixed, I left the debugging message boxes in for now just in case.

[http://litdev.co.uk/downloads/SB-Prime-TEST.zip](http://litdev.co.uk/downloads/SB-Prime-TEST.zip)

**Comment by triksterr** [comment]
Yes! This problem is solved.

But... In this test version I can't create more than one shape: the next shape removes the previous shape and its code.

**Comment by litdev1** [comment]
Oh dear - I don't see this problem so I have uploaded a test version without the debug messages to test.

[http://litdev.co.uk/downloads/SB-Prime-TEST.zip](http://litdev.co.uk/downloads/SB-Prime-TEST.zip)

**Comment by triksterr** [comment]
This test version works perfect! :)
The problem is solved!

Can You go to the next issue?

https://github.com/litdev1/SB-IDE/issues/8
